**<h1><center>Laboratorio 11: LLM y Agentes Autónomos 🤖</center></h1>**

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### **Cuerpo Docente:**

- Profesor: Ignacio Meza, Sebastian Tinoco
- Auxiliar: Catherine Benavides, Consuelo Rojas
- Ayudante: Eduardo Moya, Nicolás Ojeda

### **Equipo:**

- Nombre de alumno 1: Melanie Peña Torres
- Nombre de alumno 2: Valentina Rojas Osorio

**SUPER IMPORTANTE** - notebooks sin nombre no serán revisados

### **Link de repositorio de GitHub:** [Link](https://github.com/melaniejalea/Laboratorios-MDS7202/tree/lab11)

### **Indice**

1. [Temas a tratar](#Temas-a-tratar:)
3. [Descripcción del laboratorio](#Descripción-del-laboratorio.)
4. [Desarrollo](#Desarrollo)

## **Temas a tratar**

- Implementación de modelos de LLM y Reinforcement Learning.
- Utilización e implementación de agentes.

## **Reglas:**

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- Prohibidas las copias. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
Pueden usar cualquer material del curso que estimen conveniente.

### **Objetivos principales del laboratorio**

- Generar un modelo LLM generativo interactivo.
- Entrenar un modelo de Reinforce Learning.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

## **1. Large Language Models (4.0 puntos)**

<p align="center">
  <img src="https://trestristescriticos.com/wp-content/uploads/2021/07/telefono-gratuito-cinesur.jpg" width="350">
</p>

Joaquín no es un aficionado del cine, pero a principios de año, se propuso ver más peliculas para poder tener más temas de conversación con sus amigos y familia. Sin embargo, ya es junio y Joaquín no ha visto ninguna pelicula nueva o relevante de las que tenía en su lista y su reunión familiar bi-anual se acerca y necesita la mayor información que pueda recopilar de dichas peliculas sin tener que verlas.

Para esto, usted con su compañerx, tendrá que crear una aplicación utilizando LangChain.


---



**Instalación de librerías**

Para la creación de la aplicación, se utilizara un modelo de lenguaje (LLM) ofrecido gratuitamente por Google.

Para ello, se utilizará la API de Gemini, por lo que si no tienen acceso, se pueden crear una cuenta en el siguiente [enlace a Google AI](https://ai.google.dev/). Ahí, ir a la pestaña superior y seleccione la opción que dice ``Gemini API``.

<img src='https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/Screenshot_2024-06-13_at_12.42.32_PM.png' width='450' />

Luego, seleccione el botón que dice ``Get API key in Google AI Studio`` y hacer click en ``Crear clave de API`` para generar la llave con la que se podrá consultar al modelo de lenguaje.

<img src='https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/Screenshot_2024-06-13_at_12.45.10_PM.png?ref_type=heads' width='450' />

**Importante:** Debido a las restricciones de esta API, lo ideal es utilizar la llave a la API de manera personal.


Para mayor información sobre **LangChain**, pueden revisar la documentación en el [presente enlace](https://python.langchain.com/v0.2/docs/tutorials/summarization/ ).

In [ ]:
%%capture
!pip install langchain
!pip install langchain_google_genai
!pip install langchain-community
!pip install langchain-experimental
!pip install sentence-transformers
!pip install faiss-cpu

In [ ]:
from langchain_google_genai import ChatGoogleGenerativeAI
import os

os.environ["GOOGLE_API_KEY"] = "AIzaSyD904omnTtarZIyb9AAzjcbX8Gp1po7rMs"
llm = ChatGoogleGenerativeAI(model="gemini-1.5-flash")

### **1.1 Carga y limpieza (0.5 puntos)**

Para iniciar su titanica tarea de enseñarle a Joaquín sobre las mejores peliculas del último tiempo, tiene que revisar los script de las siguientes 3 peliculas:
* Dune 2
* Under Paris
* Joker

Debe encontrar un patrón y obtener solamente el guión de las películas. Para ello se recomienda utilizar métodos de búsqueda y reemplazo que tienen los ``string`` en Python. Adicionalmente, puede usar filtros de expresiones regulares.

Posterior a la limpieza de los guiones, debe considerar que el patrón se repite y es generalizable. ✅


In [ ]:
# Scripts de peliculas
dune2_script="https://scrapsfromtheloft.com/movies/dune-part-two-2024-transcript/"
underparis_script="https://scrapsfromtheloft.com/movies/under-paris-2024-transcript/"
joker_script="https://scrapsfromtheloft.com/movies/joker-2019-transcript/"

In [ ]:
from langchain.document_loaders import WebBaseLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

In [ ]:
def load_website_data(url):
  '''
  Escriba su código acá
  '''
  loader = WebBaseLoader(url)
  docs = loader.load()
  website_data = docs[0].page_content
  return website_data

In [ ]:
script_dune = load_website_data(dune2_script)
script_underparis = load_website_data(underparis_script)
script_joker = load_website_data(joker_script)

In [ ]:
script_dune[:500]

'\n\n\n\n\n\n\n\nDune: Part Two (2024) | Transcript - Scraps from the loft\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\r\n\t\tSkip to content\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n \n\n\n\n\n\n\n\n\n\n\n\n\n\nMOVIES\n\nMOVIE REVIEWS\nMOVIE TRANSCRIPTS\n\n\nTV SERIES\n\nTV SHOW TRANSCRIPTS\n\n\nCOMEDY\n\nSTAND-UP COMEDY TRANSCRIPTS\nGEORGE CARLIN\nDAVE CHAPPELLE\n\n\nINTERVIEWS\n\nPLAYBOY INTERVIEWS\n\n\nMUSIC\nHISTORY\nBOOKS\nOPINIONS\n \n\n Menu\n\n\nMOVIES\n\nMOVIE REVIEWS\nMOVIE TRANSCRIPTS\n\n\nTV SERIES\n\nTV SHOW TRANSCRIPTS\n\n\nCOMEDY\n\nSTAND-UP C'

In [ ]:
script_underparis[:500]

'\n\n\n\n\n\n\n\nUnder Paris (2024) | Transcript - Scraps from the loft\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\r\n\t\tSkip to content\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n \n\n\n\n\n\n\n\n\n\n\n\n\n\nMOVIES\n\nMOVIE REVIEWS\nMOVIE TRANSCRIPTS\n\n\nTV SERIES\n\nTV SHOW TRANSCRIPTS\n\n\nCOMEDY\n\nSTAND-UP COMEDY TRANSCRIPTS\nGEORGE CARLIN\nDAVE CHAPPELLE\n\n\nINTERVIEWS\n\nPLAYBOY INTERVIEWS\n\n\nMUSIC\nHISTORY\nBOOKS\nOPINIONS\n \n\n Menu\n\n\nMOVIES\n\nMOVIE REVIEWS\nMOVIE TRANSCRIPTS\n\n\nTV SERIES\n\nTV SHOW TRANSCRIPTS\n\n\nCOMEDY\n\nSTAND-UP COMED'

In [ ]:
script_joker[:500]

'\n\n\n\n\n\n\n\nJoker (2019) | Transcript - Scraps from the loft\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\r\n\t\tSkip to content\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n \n\n\n\n\n\n\n\n\n\n\n\n\n\nMOVIES\n\nMOVIE REVIEWS\nMOVIE TRANSCRIPTS\n\n\nTV SERIES\n\nTV SHOW TRANSCRIPTS\n\n\nCOMEDY\n\nSTAND-UP COMEDY TRANSCRIPTS\nGEORGE CARLIN\nDAVE CHAPPELLE\n\n\nINTERVIEWS\n\nPLAYBOY INTERVIEWS\n\n\nMUSIC\nHISTORY\nBOOKS\nOPINIONS\n \n\n Menu\n\n\nMOVIES\n\nMOVIE REVIEWS\nMOVIE TRANSCRIPTS\n\n\nTV SERIES\n\nTV SHOW TRANSCRIPTS\n\n\nCOMEDY\n\nSTAND-UP COMEDY TRA'

In [ ]:
def remove_text_before_marker(text):
    '''
    Escriba su código acá
    '''
    '''
    Notamos que hay caracteres como esto al usar la última función:
    \n[all laughing]\n[speaking in Chakobsa]
    y además hay que tener presente el * como token
    de inicio, gracias NLP!
    '''
    start_markers = ["* * *", "\n["]
    start_position = -1
    for marker in start_markers:
        position = text.find(marker)
        if position != -1:
            start_pos = position + len(marker)
            break
    if start_pos == -1:
        start_pos = 0
    text_out = text[start_pos:]
    '''
    Explorando los scripts usando la ultima función
    notamos que el script termina alrededor del token de More dado que luego
    lo que se tiene son anuncios para conseguir más scripts y rcibir anuncios
    \n[breathing heavily]\n\n\n\n\n\n\n\n\n\n\n\n\nMore:\n\nDenis Villeneuve, Dune: Part Two, Movie Transcripts \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nSHARE THIS ARTICLE \n\n\n\n\n\n\n\n \n\n\n\n\n\n
     \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nLeave a Comment Cancel ReplyYour email address will not be published
    '''
    end_marker = "\n\nMore:\n\n"
    end_position = text_out.find(end_marker)
    if end_position != -1:
        text_out = text_out[:end_position]
    text_out = text_out.strip()
    return text_out

In [ ]:
clean_dune = remove_text_before_marker(script_dune)
clean_underparis = remove_text_before_marker(script_underparis)
clean_joker = remove_text_before_marker(script_joker)

In [ ]:
clean_dune[:500]

'male voice in alien language] Power over Spice is power over all\n[suspenseful music playing]\n[Irulan] Imperial Diary. Year 10,191. Third comment. The battle for Arrakis took everyone by surprise. There were no witnesses.\n[somber music playing]\nThe Harkonnen operation was perpetrated overnight, without warning or declaration of war. By morning, the Atreides were no more. All died in the dark.\nAnd the Emperor said… nothing. Since that night, my father has not been the same. Nor have I. His inactio'

In [ ]:
clean_underparis[:500]

'[water swirling, burbling]\n“IT IS NOT THE STRONGEST OF THE SPECIES THAT SURVIVES, NOR THE MOST INTELLIGENT, IT IS THE ONE THAT IS MOST ADAPTABLE TO CHANGE.”\nBASED ON THE WORK OF CHARLES DARWIN\n[calm, mysterious music playing]\n[music builds]\n[bubbles burbling]\n[epic music intensifies]\nNORTH PACIFIC OCEAN\nMISSION OCEAN ORIGINS PROJECT “EVOLUTION”\n[woman 1] I’m here on what’s known as the “seventh continent,” a mass of refuse and garbage over 3 million kilometers in size, four to six times bigger t'

In [ ]:
clean_joker[:500]

'[news anchor on radio] The news never ends.\nThis is 1080 GCR.\nYou get all the news you need, all day long.\nGood morning.\nIt’s 42 degrees at 10:30 on this Thursday, October 15th.\nI’m Stan L. Brooks, and here’s what’s happening.\nIt’s day 18 of the garbage strike, with 10,000 tons of garbage piling up every day.\nEven the nicest sections of the city are looking like slums.\nHealth Commissioner Edward O’Rourke is declaring a city-wide state of emergency for the first time in decades.\n[O’Rourke] There’'

### **1.2 Aplicación (3.5 puntos)**

Luego de limpiar los guiones, es posible generar la aplicación deseada con el LLM. Esta aplicación tiene que ser capaz de realizar las siguientes tareas.

1. Utilizando una plantilla sobre el nombre del archivo o la URL, identifique el supuesto nombre de la película.

2. Genere un resumen en español de la película y una nota evaluativa sobre la misma. El resumen debe tener entre 3 a 5 párrafos. Además, obtener una evaluación de la película con una calificación del 1 al 10, utilizando una LLM y el contexto entregado

#### **1.2.1 Título de la película (0.5 puntos)** ✅

Para obtener el título, utilicé la siguiente plantilla:
```
 template = """
  What is the movie that appears in the description of this file or url?
  You only give me the movie name, nothing more.
  document/url: {script_path_url}
  """
```



In [ ]:
from langchain.chains import LLMChain

In [ ]:
def get_movie_title(script_path_url):
  '''
  Escriba su código acá
  '''
  template = """
  What is the movie that appears in the description of this file or url?
  You only give me the movie name, nothing more.
  document/url: {script_path_url}
  """
  prompt_template = PromptTemplate.from_template(template)
  prompt_chain = LLMChain(llm=llm, prompt=prompt_template)
  result = prompt_chain.invoke({'script_path_url':script_path_url})
  return result['text'].strip()

In [ ]:
script1_title = get_movie_title(dune2_script)
script1_title

'Dune: Part Two'

In [ ]:
script2_title = get_movie_title(underparis_script)
script2_title

'Under Paris'

In [ ]:
script3_title = get_movie_title(joker_script)
script3_title

'Joker'

#### **1.2.2 Resumen (1.0 puntos)** ✅

Como se vió en clases, las LLM no pueden manejar cadenas de texto muy largas, esto es debido a que, dependiendo de su naturaleza, solo manejan ventanas de contexto que estan asociadas a caracteristicas de la red y del entrenamiento utilizado.

Por ello, es altamente importante que si se desea hacer un resumen del texto, este se haga realizando un tipo de map/reduce sobre el texto. De manera que en cada una de las iteraciones se vaya disminuyendo el tamaño del texto, pero hay que tener cuidado con que le modelo vaya guardando el contexto de escenas previas.

In [ ]:
from langchain.chains import MapReduceDocumentsChain, ReduceDocumentsChain
from langchain.chains import StuffDocumentsChain, LLMChain
from langchain.chains.combine_documents.stuff import StuffDocumentsChain
from langchain_core.prompts import PromptTemplate
from langchain_text_splitters import RecursiveCharacterTextSplitter

In [ ]:
#No cambiar función

def map_reduce_text(script, map_template, reduce_template):

  # Map
  """
  map_prompt, crear el prompt desde el template
  map_chain, crear la cadena desde el prompt
  """
  map_prompt = PromptTemplate.from_template(map_template)
  map_chain = LLMChain(llm=llm, prompt=map_prompt)
  # Reduce
  """
  reduce_prompt, crear el prompt desde el template
  reduce_chain, crear la cadena desde el prompt
  """
  reduce_prompt = PromptTemplate.from_template(reduce_template)
  reduce_chain = LLMChain(llm=llm, prompt=reduce_prompt)
  # Combine
  """
  Combinar y reducir los documentos, utilizar StuffDocumentsChain
  y ReduceDocuentsChain con un máximo de 4000 tokens
  """
  combine_documents_chain = StuffDocumentsChain(
    llm_chain=reduce_chain, document_variable_name="script")

  reduce_documents_chain = ReduceDocumentsChain(
    combine_documents_chain=combine_documents_chain,
    collapse_documents_chain=combine_documents_chain,
    token_max=4000,)

  # Map/Reduce
  """
  Uilizar MapReduceDocumentsChain
  """
  map_reduce_chain = MapReduceDocumentsChain(
    llm_chain=map_chain,
    reduce_documents_chain=reduce_documents_chain,
    document_variable_name="script",
    return_intermediate_steps=False,)


  # Text splitter
  """
  Usar RecursiveCharacterTextSplitter
  """
  text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    chunk_size=1000, chunk_overlap=0)

  # Resultado
  split_script = text_splitter.create_documents([script])
  result = map_reduce_chain.invoke(split_script)

  return result["output_text"]

In [ ]:
%%capture
!pip install tiktoken

In [ ]:
# crear templates

map_template_summary = """The following is a script from a movie
{script}
Based on this script, please identify the main themes from it.
Helpful Answer:"""

reduce_template_summary ="""The following is a script from a movie:
{script}
Take it and distill it into a final, consolidated summary of the main themes.
Name the movie and then include it's summary.This summary must be in between 3 or 5 paragraphs.
It also needs to include a review for this movie rating it from a 1 to a 10, 10 being the highest score.
Helpful Answer:"""

In [ ]:
# imprimir resumenes de películas.
answer_summary = map_reduce_text(clean_dune, map_template_summary, reduce_template_summary)
print(answer_summary)

## Dune (2021)

"Dune" is a sprawling epic that plunges viewers into a universe of political intrigue, ecological wonder, and spiritual awakening.  The film follows Paul Atreides, a young noble thrust into the center of a galactic power struggle over the control of a precious resource known as "spice."  Forced to flee his home after a betrayal, Paul finds refuge among the Fremen, a fierce desert people who hold the key to his survival and destiny. 

As Paul adapts to the harsh environment and embraces the Fremen culture, he begins to understand the power of belief and the weight of his own prophecy. The film masterfully explores themes of power, destiny, and faith. Paul grapples with his prophetic abilities, the weight of his destiny, and the influence of the Bene Gesserit, a powerful organization that has been manipulating events for generations. The Fremen, with their deep reverence for the desert and their unwavering belief in Paul's prophesied role, represent the power of faith and

In [ ]:
answer_summary2 = map_reduce_text(clean_underparis, map_template_summary, reduce_template_summary)
print(answer_summary2)

## Evolution: A Dive into Environmental Crisis and Human Resilience

"Evolution" is a captivating film that takes viewers on a journey through a world dramatically altered by human impact on the environment.  The film opens with a stark portrayal of the "seventh continent," a massive plastic island in the Pacific Ocean, setting the stage for a narrative that explores the consequences of pollution and the struggle for survival in a rapidly changing world.  The story centers around a team of scientists dedicated to understanding the ocean's mysteries, but their research is conducted within an environment heavily impacted by human activity.  Through their encounters with sharks, creatures known for their resilience and ability to thrive in challenging environments, the film explores the themes of adaptation and survival in the face of adversity. 

"Evolution" offers a thought-provoking exploration of humanity's relationship with nature, highlighting both the beauty and the fragility of th

Por alguna razón por más que pidamos esta película "Under Paris" nos entrega otras películas, asumimos por que se estreno este año, por lo que el modelo no funciona bien para él. Fecha de estreno inicial de Under Paris: 5 de junio de 2024

In [ ]:
answer_summary3 = map_reduce_text(clean_joker, map_template_summary, reduce_template_summary)
print(answer_summary3)

## Joker (2019)

"Joker" is a chilling and thought-provoking exploration of mental illness, social inequality, and the potential for violence when individuals feel ignored and marginalized.  The film follows Arthur Fleck, a struggling clown with a mental health condition, as he navigates a world that seems determined to crush him. Arthur's descent into madness is fueled by a combination of societal neglect, a lack of support, and a growing sense of alienation. The film powerfully depicts the devastating consequences of mental illness when individuals are denied the resources and compassion they need. 

Beyond the individual struggles of Arthur, "Joker" also paints a bleak picture of a society on the brink of collapse. The film highlights the stark divide between the wealthy elite and the struggling lower class, showcasing the corruption of power and the hypocrisy of those who hold it.  Arthur's resentment towards the "pricks" in Gotham City, particularly Thomas Wayne, a symbol of privi

Adicionalmente, Joaquín sabe que su primo favorito le gusta ``Dune: Part 2`` por lo que le gustaría tener mayor información al respecto, para ello realice las siguientes tareas:


3. Genere un gráfico que muestre los personajes de la película con más apariciones en la misma. ✅
4. Genere una tabla en pandas con los 3 personajes que más aparecen, indicando el nombre del actor y su edad actual más uno (ojo edad + 1).
5. Cree una función que responda preguntas sobre la película basándose en la información del texto entregado (OJO: las preguntas y salidas deben ser en español). Luego, responda las siguientes preguntas:
* ¿Qué y quién es Lisan al-Gaib?
* ¿Qué personaje no cree en la profecía pero es parte de ella?
* ¿Cuál es el objetivo de Feyd-Rautha?
6. Utilizando el top 3 de personajes que más aparecen en la película, genere con el modelo LLM y utilizando el contexto del guion, las 6 estadísticas que demuestren las habilidades de los personajes: Intelligence, Strength, Charisma, Wisdom, Emotional Resilience, y Creativity.

#### **1.2.3 Personajes (0.5 puntos)**

En la siguiente sección, tiene que entregar un template de personajes y redicción

In [ ]:
map_template_characters = """The following is a movie script:
{script}
Identify all characters mentioned in the script.
Characters appear in the format "[name_character]" within the text.
Ignore every other name that does not follow this format.
Do not add any additional information to get the names, like sources
from books or comics of the same name.
Helpful Answer:
"""

reduce_template_characters = """Here are summaries from the script:
{script}
From these summaries, provide a final list in Python list format stored in a variable named 'characters_name'.
The list must only contain names. The list must not contain text characters like '\n' and of the sort.
Helpful Answer:
"""

answer_character_list = map_reduce_text(
    clean_dune,
    map_template_characters,
    reduce_template_characters
)

In [ ]:
answer_character_list

"```python\ncharacters_name = ['Irulan', 'Paul', 'Jessica', 'Stilgar', 'Chani', 'Harkonnen soldier', 'Harkonnen commander', 'Rabban', 'Fremen sentinel', 'sentinel leader', 'Jamis', 'Shishakli', 'old watermaster', 'Fedaykin fighter', 'Usul', 'Muad’Dib', 'Baron Harkonnen', 'male Fedaykin fighter', 'nun', 'Emperor', 'Reverend Mother Mohiam', 'Feyd-Rautha Harkonnen', 'Feyd-Rautha', 'weapon master', 'attendant', 'slave master', 'Lanville', 'Lady Margot Fenring', 'Bene Gesserit sister 1', 'Bene Gesserit sister 2', 'gladiator arena announcer', 'Baron Harkonnen', 'Rabban', 'Gurney', 'Maker Keeper', 'Alia', 'Rabban Harkonnen', 'Paul Muad’Dib Atreides', 'Bashar']\n```"

In [ ]:
import pandas as pd
import plotly.express as px
from itertools import count
import ast
import re
import numpy as np

def plot_characters(answer_character_list):
  # Clear answer
  characters_list = re.findall(r"'(.*?)'", answer_character_list)

  # distil the characters output
  """
  Recomendación, utilizar un diccionario para ordenar los personajes
  """
  count_characters = {}
  for character in characters_list:
    pattern = rf'\[{re.escape(character)}\]'
    matches = re.findall(pattern, clean_dune)
    count_characters[character] = len(matches)

  # Create dataframe
  """
  De diccionario a DataFrame
  """
  df_characters = pd.DataFrame(list(count_characters.items()), columns=['Personaje', 'Frecuencia en el script'])
  df_characters = df_characters.sort_values(by='Frecuencia en el script', ascending=False)

  # Graficar datos

  fig = px.bar(
      df_characters,
      x='Personaje',
      y='Frecuencia en el script',
      title='Personajes de la película y su frecuencia',
      template='plotly_white'
  )
  fig.show()

  #Retornar los personajes
  return df_characters[['Personaje']].reset_index(drop=True)

In [ ]:
df = plot_characters(answer_character_list)

#### **1.2.4 Actores principales (0.75 puntos)**

Importante saber que el script **no** maneja información de los actores, por ello, es importante que nuestra LLM tenga acceso a internet, de manera de poder realizar búsquedas que nos ayuden a completar la información consultada.

Para esto, utilizaremos agentes combinados con react para realzar la consulta y asegurarnos de que la respuesta es correcta.

In [ ]:
from langchain.agents import load_tools
from langchain.agents import AgentType, initialize_agent

In [ ]:
# Key para realizar una busqueda
os.environ["SERPER_API_KEY"] = 'd63e62662ef63eb9e44ab133d191f7a99a0024a3'

In [ ]:
personajes = df.Personaje.to_list()

In [ ]:
def get_actors_and_age(character):

  # Inicializar tools y agente.
  tools = load_tools(["google-serper", "llm-math"], llm=llm)
  agent = initialize_agent(tools, llm, agent="zero-shot-react-description", verbose=True)

  # Crear template de query
  query_template = """  Here is a list of characters from the movie {movie}: {characters}
    For each character, find the name of the actor who played them and their current age plus one year (age + 1).
    Provide the results in the following format:
    Character: [character name], Actor: [actor name], Age+1: [actor age + 1].
  """
  # Crear prompt y usar agente para la búsqueda.
  reduce_prompt = PromptTemplate.from_template(query_template)

  # Retornar Nombre y Edad + 1
  reduce_prompt = PromptTemplate.from_template(query_template)
  agent_templated = reduce_prompt | agent
  agent_answer = agent_templated.invoke(
      {
          'movie': script1_title,
          'characters': personajes
      }
  )
  return agent_answer['output']

In [ ]:
get_actors_and_age(personajes)



> Entering new AgentExecutor chain...
Thought: I need to find the actors and their ages for each character in the list. I can use Google Search to find this information.
Action: google_serper
Action Input: "Dune Part Two cast and age"
Observation: Science fiction epic that serves as a continuation of 2021's Dune. Paul Atreides reunites with the Fremen as he seeks revenge against those who destroyed his ... Timothée Chalamet · Paul Atreides ; Zendaya · Chani ; Rebecca Ferguson · Jessica ; Javier Bardem · Stilgar ; Josh Brolin · Gurney Halleck. Your guide to the full cast and characters of Dune: Part Two which is led by Timothée Chalamet as Paul Atreides and Zendaya as Chani ... Dune: Part Two (2024) cast and crew credits, including actors, actresses, directors, writers and more. Missing: age | Show results with:age. Share on Social Media · Timothée Chalamet as Paul Atreides · Zendaya as Chani · Javier Bardem as Stilgar · Rebecca Ferguson as Lady Jessica · Josh ... The Dune 2 cast feat

KeyboardInterrupt: 

**Explicar metodología utilizada**

:

#### **1.2.5 Personajes Stats (0.5 puntos)**

Esta parte es similar al punto 2. La clave esta en crear un buen prompting que nos permita generar las estadísticas basandonos en una búsqueda por map/reduce.

Tras la búsqueda, la idea es tener una función de Python que nos permita generar el gráfico deseado y tener el resumen de los personajes.


In [ ]:
import json

def map_reduce_text(script, character):
  # Map
  map_template = """
  Crear template, utilizar las palabras claves:
  Intelligence, Charisma, Strength, Wisdom, Emotional Resilience and Creativity.
  """

  # crear prompt y cadena
  map_template += template_complemt
  map_prompt = # ...
  map_chain = # ...

  # Reduce
  reduce_template = """
  Crear prompt de reducción.
  Reducir, dado el perfil, en escala del 1 al 10 las cualidades mencionadas
  """
  reduce_prompt = # ...
  reduce_chain = # ...

  # Reduce
  """
  Reducir y combinar los documentos con un máximo de 4000 tokens
  """

  # Map/Reduce
  """
  Uilizar MapReduceDocumentsChain
  """


  # Text splitter
  """
  Usar RecursiveCharacterTextSplitter
  """

  result = map_reduce_chain.invoke(split_script)
  return result["output_text"]


# Formato del perfil
def format_profile(answer_character_profile):
  """
  Crear un json con las caracteristicas y que retorne
  (final_profile, stats) del personaje
  """
  return (final_profile, stats)


In [ ]:
# Escriba su respuesta acá

In [ ]:
final_profile, stats = format_profile(answer_character_profile)

In [ ]:
print(final_profile)

In [ ]:
# Función para gráficar stats. No Tocar.

import plotly.graph_objects as go
import numpy as np

def plot_stats(stats, character_name="Paul Atreides"):
    base_stats = [
        "Intelligence", "Charisma", "Strength",
        "Wisdom", "Emotional Resilience", "Creativity"
    ]
    for stat in base_stats:
        if stat not in stats:
            stats[stat] = 0

    labels = list(stats.keys())
    stats_values = list(stats.values())
    stats_values += stats_values[:1]
    labels += labels[:1]

    # Plotly figure
    fig = go.Figure()
    fig.add_trace(go.Scatterpolar(
        r=stats_values,
        theta=labels,
        fill='toself',
        name=character_name
    ))

    fig.update_layout(
        polar=dict(
            radialaxis=dict(
                visible=True,
                range=[0, max(stats_values)]
            )
        ),
        showlegend=False,
        title=character_name
    )

    return fig

fig = plot_stats(stats)

#### **Comentar (0.25 puntos)**
Explicar metodología y secuencia lógica de cada una de las respuestas. Además responda:

* ¿Qué otras tareas se podría realizar? De dos ejemplos con la metodología asociada.

* ¿Cual es la importancia de los prompt y como estos afectan al desempeño de los LLM?

* ¿Alguna de sus respuestas fue una 'alucinación'? ¿Por qué sucede esto?

## **2. Reinforcement Learning (2.0 puntos)**

En esta sección van a usar métodos de RL para resolver dos problemas interesantes: `Blackjack` y `LunarLander`.

In [1]:
!pip install -qqq gymnasium stable_baselines3
!pip install -qqq swig
!pip install -qqq gymnasium[box2d]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 953.9/953.9 kB 4.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.3/182.3 kB 9.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21.3 MB 72.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 21.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 374.4/374.4 kB 10.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


### **2.1 Blackjack (1.0 puntos)**

<p align="center">
  <img src="https://www.recreoviral.com/wp-content/uploads/2016/08/s3.amazonaws.com-Math.gif"
" width="400">
</p>

Joaquín es fanático del Blackjack, por lo que en esta subsección implementarán métodos de RL y así generar una estrategia para que pueda ~~ir al casino a  hacerse millonario~~ aprender a resolver problemas mediante RL.

Comencemos primero preparando el ambiente. El siguiente bloque de código transforma las observaciones del ambiente a `np.array`:


In [2]:
import gymnasium as gym
from gymnasium.spaces import MultiDiscrete
import numpy as np

class FlattenObservation(gym.ObservationWrapper):
    def __init__(self, env):
        super(FlattenObservation, self).__init__(env)
        self.observation_space = MultiDiscrete(np.array([32, 11, 2]))

    def observation(self, observation):
        return np.array(observation).flatten()

# Create and wrap the environment
env = gym.make("Blackjack-v1")
env = FlattenObservation(env)

#### **2.1.1 Descripción de MDP (0.2 puntos)**

Entregue una breve descripción sobre el ambiente [Blackjack](https://gymnasium.farama.org/environments/toy_text/blackjack/) y su formulación en MDP, distinguiendo de forma clara y concisa los estados, acciones y recompensas.

**Estados**
1. Suma actual del jugador (entero)
2. Valor de la carta visible del crupier (1-10, donde 1 representa un as)
3. Si el jugador tiene un as usable (0 o 1)

**Acciones:** El espacio de acciones es discreto con dos posibilidades
- 0: Plantarse (dejar de recibir cartas)
- 1: Pedir (solicitar una carta adicional)

**Recompensas:** La función de recompensa se define como
- Ganar: +1
- Perder: -1
- Empate: 0
- Ganar con un blackjack natural:
  - +1.5 si natural=True
  - +1 si natural=False

#### **2.1.2 Generando un Baseline (0.2 puntos)**

* Simule un escenario en donde se escojan acciones aleatorias. Repita esta
simulación 5000 veces y reporte el promedio y desviación de las recompensas.
* ¿Cómo calificaría el performance de esta política?
* ¿Cómo podría interpretar las recompensas obtenidas?

In [3]:
import numpy as np

episodes = 5000
rewards = np.zeros(episodes)

for episode in range(episodes):
    obs, info = env.reset()
    done = False

    while not done:
        action = env.action_space.sample()
        obs, reward, done, _, info = env.step(action)

    rewards[episode] = reward

mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print("Mean reward:", mean_reward)
print("Standard deviation of rewards:", std_reward)

Mean reward: -0.416
Standard deviation of rewards: 0.8859706541415465



**¿Cómo calificaría el performance de esta política?**

La política de seleccionar acciones aleatorias en el Blackjack es **deficiente**. La recompensa promedio es negativa (-0.416), indicando pérdidas en promedio. La alta desviación estándar (0.886) muestra una gran variabilidad en las recompensas, esperable por las decisiones aleatorias.

**¿Cómo podría interpretar las recompensas obtenidas?**

- **Recompensa promedio negativa** (-0.416): Esto significa que en promedio, el jugador termina perdiendo dinero, si uno sigue esta estrategia a largo plazo, es más probable que termine con menos dinero del que empezó.
- **Alta desviación estándar** (0.886): Esto muestra que las recompensas pueden variar mucho de una ronda a otra, a veces se puede ganar un poco más, pero también se puede perder bastante.


#### **2.1.3 Entrenamiento de modelo (0.2 puntos)**

A partir del siguiente [enlace](https://stable-baselines3.readthedocs.io/en/master/guide/algos.html), escoja un modelo de `stable_baselines3` y entrenelo para resolver el ambiente `Blackjack`.

In [4]:
from stable_baselines3 import DQN
from stable_baselines3.common.evaluation import evaluate_policy

# Instantiate the agent
model = DQN("MlpPolicy", env, verbose=0)

# Train the agent and display a progress bar
model.learn(total_timesteps=int(2e5), progress_bar=True)

Output()

#### **2.1.4 Evaluación de modelo (0.2 puntos)**

* Repita el ejercicio 2.1.2 pero utilizando el modelo entrenado.
* ¿Cómo es el performance de su agente?
* ¿Es mejor o peor que el escenario baseline?

In [5]:
from stable_baselines3.common.evaluation import evaluate_policy

mean_reward, std_reward = evaluate_policy(model, env, n_eval_episodes = 5_000)
mean_reward, std_reward

/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/evaluation.py:67: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  warnings.warn(


(-0.0482, 0.9498824979964626)

Para evaluar el desempeño del agente entrenado con el modelo de DQN en el problema de Blackjack, consideramos:

- Promedio de recompensas: -0.0482
- Desviación estándar de las recompensas: 0.9499

**Desempeño del Agente**

El agente entrenado muestra una recompensa promedio cercana a cero (-0.0482), indicando resultados neutros en promedio. Sin embargo, la alta desviación estándar (0.9498) revela que las recompensas varían mucho entre simulaciones, debido a la naturaleza aleatoria del juego y las decisiones tomadas por el agente.

**Comparación con el Baseline**

El agente entrenado muestra un desempeño ligeramente mejor, con un promedio de recompensas más cercano a cero, lo que sugiere una estrategia más informada que podría estar minimizando las pérdidas promedio. Sin embargo, la alta variabilidad en las recompensas indica que aún hay espacio para mejorar la consistencia y la efectividad del agente en el juego de Blackjack.


#### **2.1.5 Estudio de acciones (0.2 puntos)**

* Genere una función que reciba un estado y retorne la accion del agente.
* Luego, use esta función para entregar la acción escogida frente a los siguientes escenarios:

  * Suma de cartas del agente es 6, dealer muestra un 7, agente no tiene tiene un as
  * Suma de cartas del agente es 19, dealer muestra un 3, agente tiene tiene un as

* ¿Son coherentes sus acciones con las reglas del juego?

Hint: ¿A que clase de python pertenecen los estados? Pruebe a usar el método `.reset` para saberlo.

In [6]:
state = env.reset()
type(state)

tuple

In [7]:
predicted_action, prediction_info = model.predict(np.array([6, 7, 0]), deterministic=True)
predicted_action

array(1)

El modelo predice array(1), lo que significa que el agente debería pedir otra carta.

In [8]:
predicted_action, prediction_info = model.predict(np.array([19, 3, 1]), deterministic=True)
predicted_action

array(0)

El modelo predice array(0), lo que significa que el agente no debería pedir otra carta.

Recordemos las reglas básicas para evaluar la coherencia:
- El agente debe pedir otra carta si su suma de cartas es menor a 21 y no tiene un blackjack natural.
- El agente puede *plantarse* si considera que su suma de cartas es lo suficientemente alta para ganar sin exceder 21. personalmente creo que desde 17 en adelante es weno :-)

**Conclusiones**:
- **Escenario 1**, el agente tiene una suma baja (6) frente a una carta fuerte del dealer (7), por lo tanto es coherente que el modelo sugiera pedir otra carta.
- **Escenario 2**, el agente ya tiene una suma alta (19) y tiene un as, lo cual le da flexibilidad para decidir plantarse sin riesgo de pasarse de 21.

Ambas acciones predichas por el modelo son coherentes con las reglas del juego según los escenarios presentados.

### **2.2 LunarLander**

<p align="center">
  <img src="https://i.redd.it/097t6tk29zf51.jpg"
" width="400">
</p>

Similar a la sección 2.1, en esta sección usted se encargará de implementar una gente de RL que pueda resolver el ambiente `LunarLander`.


#### **2.2.1 Descripción de MDP (0.2 puntos)**


Comencemos preparando el ambiente:

In [9]:
import gymnasium as gym
env = gym.make("LunarLander-v2", render_mode = "rgb_array", continuous = True) # notar el parámetro continuous = True

* Entregue una breve descripción sobre el ambiente [LunarLander](https://gymnasium.farama.org/environments/box2d/lunar_lander/) y su formulación en MDP, distinguiendo de forma clara y concisa los estados, acciones y recompensas.
* ¿Como se distinguen las acciones de este ambiente en comparación a `Blackjack`?
* En la preparación del ambiente se especifica el parámetro `continuous = True`. ¿Que implicancias tiene esto sobre el ambiente?

El MDP para Lunar Lander con `continuous=True` se puede describir de la siguiente manera:

**Estados**: Vector de 8 dimensiones que contiene:
  1. Coordenada x del módulo lunar
  2. Coordenada y del módulo lunar
  3. Velocidad lineal en x
  4. Velocidad lineal en y
  5. Ángulo del módulo lunar
  6. Velocidad angular
  7. Booleano indicando si la pierna izquierda está en contacto con el suelo
  8. Booleano indicando si la pierna derecha está en contacto con el suelo

**Acciones**: Espacio continuo representado como `Box(-1, +1, (2,), dtype=np.float32)`:
  - Primera coordenada: Controla el acelerador del motor principal (0-100% de potencia)
  - Segunda coordenada: Controla los propulsores laterales

**Recompensas**:
  - Aumenta cuanto más cerca está el módulo de la plataforma de aterrizaje
  - Aumenta cuanto más lento se mueve el módulo
  - Disminuye cuanto más inclinado está el módulo
  - +10 puntos por cada pierna en contacto con el suelo
  - -0.03 puntos por cada frame que un motor lateral está encendido
  - -0.3 puntos por cada frame que el motor principal está encendido
  - -100 puntos por estrellarse, +100 puntos por aterrizar con seguridad

Las acciones en Lunar Lander son continuas, a diferencia de Blackjack donde eran discretas (pedir o plantarse). Esto permite un control más preciso, pero también hace que el espacio de acciones sea más complejo.

#### **2.2.2 Generando un Baseline (0.2 puntos)**

* Simule un escenario en donde se escojan acciones aleatorias. Repita esta simulación 10 veces y reporte el promedio y desviación de las recompensas.
* ¿Cómo calificaría el performance de esta política?

 Este performance es tremendamente pobre. Una recompensa constante de -100 sugiere que el módulo se está estrellando en cada episodio, lo cual es el peor resultado posible, pero esperado para acciones aleatorias en un ambiente tan complejo.

In [10]:
import numpy as np

episodes = 10
rewards = np.zeros(episodes)

for episode in range(episodes):
    obs, info = env.reset()
    done = False

    while not done:
        action = env.action_space.sample()
        obs, reward, done, _, info = env.step(action)

    rewards[episode] = reward

mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print("Mean reward:", mean_reward)
print("Standard deviation of rewards:", std_reward)


Mean reward: -100.0
Standard deviation of rewards: 0.0


#### **2.2.3 Entrenamiento de modelo (0.2 puntos)**

* A partir del siguiente [enlace](https://stable-baselines3.readthedocs.io/en/master/guide/algos.html), escoja un modelo de `stable_baselines3` y entrenelo para resolver el ambiente `LunarLander` **usando 10000 timesteps de entrenamiento**.

In [11]:
from stable_baselines3 import SAC
from stable_baselines3.common.evaluation import evaluate_policy

# Instantiate the agent
model = SAC("MlpPolicy", env, verbose=0)

# Train the agent and display a progress bar
model.learn(total_timesteps=int(1e4), progress_bar=True)

Output()

#### **2.2.4 Evaluación de modelo (0.2 puntos)**

* Repita el ejercicio 2.2.2 pero utilizando el modelo entrenado.
* ¿Cómo es el performance de su agente? ¿Es mejor o peor que el escenario baseline?

In [12]:
from stable_baselines3.common.evaluation import evaluate_policy

mean_reward, std_reward = evaluate_policy(model, env, n_eval_episodes = 50)
mean_reward, std_reward

(-245.62136531946098, 163.17298812826402)

El algoritmo SAC muestra un rendimiento promedio de -245.621, considerablemente peor que el baseline de -100 y un desempeño inconsistente con alta variabilidad (desviación estándar de 163.172 frente a 0 del baseline). Este resultado sugiere que el agente SAC podría estar tomando acciones poco inteligentes en lugar de mejorar la situación, posiblemente debido a malos ajustes en los hiperparámetros.

#### **2.2.5 Optimización de modelo (0.2 puntos)**

* Repita los ejercicios 2.2.3 y 2.2.4 hasta obtener un nivel de recompensas promedio mayor a 50. Para esto, puede cambiar manualmente parámetros como:
  - `total_timesteps`
  - `learning_rate`
  - `batch_size`

* Una vez optimizado el modelo, use la función `export_gif` entregada para estudiar el comportamiento de su agente en la resolución del ambiente, comente sobre sus resultados.

* Adjunte el gif generado en su entrega. Si, además, adjuntan el gif en el markdown tendrán un bonus de 0.1.

Se demoro mucho en correo y nos explotó el pc a ambas. NO CORRER, GRANADA! 🚀🔥🧨🤯

In [ ]:
from stable_baselines3 import SAC
from stable_baselines3.common.evaluation import evaluate_policy

# Instantiate the agent
model = SAC("MlpPolicy", env, verbose=0)

# Train the agent and display a progress bar
model.learn(total_timesteps=int(2e5), progress_bar=True)

mean_reward, std_reward = evaluate_policy(model, env, n_eval_episodes = 10)
mean_reward, std_reward

In [ ]:
import imageio
import numpy as np

def export_gif(model, n = 5):
  '''
  función que exporta a gif el comportamiento del agente en n episodios
  '''
  images = []
  for episode in range(n):
    obs = model.env.reset()
    img = model.env.render()
    done = False
    while not done:
      images.append(img)
      action, _ = model.predict(obs)
      obs, reward, done, info = model.env.step(action)
      img = model.env.render(mode="rgb_array")

  imageio.mimsave("agent_performance.gif", [np.array(img) for i, img in enumerate(images) if i%2 == 0], fps=29)

In [ ]:
# escriba su respuesta acá